server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -310,7 +310,12 @@ public boolean hasPassword() {
     private Cipher createCipher(int opmode, char[] password, byte[] salt, byte[] iv) throws GeneralSecurityException {
         PBEKeySpec keySpec = new PBEKeySpec(password, salt, KDF_ITERS, CIPHER_KEY_BITS);
         SecretKeyFactory keyFactory = SecretKeyFactory.getInstance(KDF_ALGO);
-        SecretKey secretKey = keyFactory.generateSecret(keySpec);
+        SecretKey secretKey;
+        try {
+            secretKey = keyFactory.generateSecret(keySpec);
+        } catch (AssertionError ae) {
+            throw new GeneralSecurityException("Error generating an encryption key from the provided password. ", ae);
```

Comment:
We throw `ElasticsearchException` in other places while throwing `GeneralSecurityException` here. Do we want to be consistent?
Nit: extra space (and the period?) at the end of the error message string :)

Reply:
> We throw ElasticsearchException in other places while throwing GeneralSecurityException here. Do we want to be consistent?

We don't have consistency over this. I don't know if it should. Choice might seem arbitrary but I did put some thought in it.  I went with a GeneralSecurityException in the KeyStoreWrapper since it is an exception that is already used in similar issues here. I chose ElasticsearchException in Hasher as a generic elasticsearch related exception because I didn't find any one more specific that would make much sense. Happy to discuss and change that approach if you or Lyudmila have better suggestions